mechanism to increase gas cost for each message emitted within one execution

alsrdn opened the pull request:
Implement a mechanism to increase gas cost for each message emitted within one execution.
Also fix a few test issues and add more tests for contract level messages.

alsrdn committed:
ee/contract_messages: add more tests

Add test for checking that messages topics are carried over to an
upgraded contract.
Fix check for not allowing session code to register message topics and
add test.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
types/chainspec: fix `WasmConfig` serialization for message limits

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
types/key: fix `key_max_serialized_length` test

Now the `Message` key has the largest serialized length so the test
needs to be updated.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
types/key: fix display test for `Key::Message` variants

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
types/contract_messages: fix `MessageAddr` from string parsing

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
types: fix lint errors caused by the import of `Key` for contract messages

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
ee/contract_messages: increase gas cost for each emitted message

Implemented a mechanism to increase the gas cost for emitting messages
linearly for each execution.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
ee/tests: adjust expected costs for faucet host fn metrics

Adjust the expected costs for the faucet and the host function
metrics contract since the code size has increased due to the addition
of the contract messages code. Also the size of the addressable entity
stored value has increased because of the stored topic names.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
casper-validation: fix fixture for `AddressableEntity`

The `AddressableEntity` record now includes the name of the message
topics so the fixture needs to be adjusted.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
ee/contract_messages: return error if topic is full

When emitting messages, return an error if there's no space left in the
topic.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

rafal-ch commented on execution_engine/src/runtime/externals.rs:
```diff
@@ -1184,6 +1184,9 @@ where
 
                 let result = self.emit_message(topic_name, message)?;
+                if result.is_ok() {
+                    self.charge_emit_message()?;
```
Are these calls immediate? I.e. if we fail to charge, do we still emit the message?

alsrdn replied:
The effects should be reverted if we fail to charge since we would trap with a `GasLimit` error.
What I wanted to do is charge if the message was actually emitted. If an `ApiError` occurred, we shouldn't charge for it. Are you suggesting we should charge every attempt to emit a message? Then the next attempt would have increased gas or have the previous gas cost?

mpapierski replied:
As discussed on slack we should always charge, but increase the counter if the message is actually emitted - in this case, the user has a chance to retry the message by (for instance) shrinking the message to fit the limit.

alsrdn replied:
@mpapierski @rafal-ch changed the behaviour here: https://github.com/casper-network/casper-node/pull/4344/commits/8171806b24ac2178a16b6721587d073eeb42a0d5

Also got rid of the `MessageCosts` config and added the parameter for cost increase to `HostFunctionCosts` which I think is more cleaner.

Now within an execution, the first call to emit a message will charge `host_function_cost.emit_message.cost()`.  The second call to emit will have cost `host_function_cost.emit_message.cost() + host_function_cost.cost_increase_per_message`. If for the second call a message was not emitted due to an `ApiError`, the third call will cost the same as the previous one (`host_function_cost.emit_message.cost() + host_function_cost.cost_increase_per_message`).

rafal-ch commented on execution_engine_testing/tests/src/test/contract_messages.rs:
```diff
@@ -494,3 +542,241 @@ fn should_not_exceed_configured_limits() {
+    const EMIT_MULTIPLE_EXPECTED_COST: u32 = FIRST_MESSAGE_EMIT_COST * MESSAGES_TO_EMIT
+        + (MESSAGES_TO_EMIT - 1) * MESSAGES_TO_EMIT / 2 * COST_INCREASE_PER_MESSAGE;
+    const EMIT_MESSAGES_FROM_MULTIPLE_CONTRACTS: u32 =
+        FIRST_MESSAGE_EMIT_COST * 3 + 3 * COST_INCREASE_PER_MESSAGE; // simplification of above
```
`EMIT_MULTIPLE_EXPECTED_COST` is 700, but `EMIT_MESSAGES_FROM_MULTIPLE_CONTRACTS` is 450.
Simplification suggests that the values stay the same, but formula is simpler.

Also, can we name the magic number `3`?

alsrdn replied:
No, the formula is the same, the expression is simplified when `MESSAGES_TO_EMIT == 3`. We can add a `const fn` with the formula and get rid of these expression. WDYT?

alsrdn replied:
Changed here: https://github.com/casper-network/casper-node/pull/4344/commits/ffe7e18b9a7139defaa1cebf8c97bde61722a229

rafal-ch replied:
Thanks, it's more clear this way :+1:

rafal-ch commented on execution_engine_testing/tests/src/test/contract_messages.rs:
```diff
@@ -494,3 +542,241 @@ fn should_not_exceed_configured_limits() {
+    let emit_message_gas_cost = builder.borrow().last_exec_gas_cost().value();
+    assert_eq!(emit_message_gas_cost, FIRST_MESSAGE_EMIT_COST.into());
+
+    // Check gas cost when multiple messages are emitted from different contracts.
```
Maybe consider moving this test-case to a separate test?

alsrdn replied:
Moving it will add a lot of boiler plate which I don't think it's warranted in this situation. I would prefer to leave the check in this test which is related to increasing gas costs.

rafal-ch replied:
Ok, fair enough.
Please just confirm if my understanding of this test scenario is correct:
```
Step 1:
1. Emit one message via "emit_message" entry point in the original contract
2. Expect: FIRST_MESSAGE_EMIT_COST

Step 2:
1. Emit 4 messages from single contract via "emit_multiple_messages" entry point in the original contract
2. Expect: emit_cost_per_execution(4)

Step 3:
1. Again emit one message via "emit_message" entry point in the original contract (to assert that costs calculation is reset across executions)
2. Expect: FIRST_MESSAGE_EMIT_COST

Step 4:
1. Emit 3 messages 
    - two via "emit_message_from_each_version" entry point in the upgraded contract
    - one via "emit_message" entry point in the original contract
2. Expect: emit_cost_per_execution(3)
```

alsrdn replied:
yes, this is the expected scenario

rafal-ch commented on types/src/chainspec/vm_config/host_function_costs.rs:
```diff
@@ -301,6 +316,125 @@ pub struct HostFunctionCosts {
+    }
+
+    fn is_zero(&self) -> bool {
+        todo!()
```
left on purpose for next PRs?

alsrdn replied:
No, actually forgot about it. Good catch!

alsrdn replied:
Added here: https://github.com/casper-network/casper-node/pull/4344/commits/ffe7e18b9a7139defaa1cebf8c97bde61722a229

rafal-ch commented on types/src/chainspec/vm_config/host_function_costs.rs:
```diff
@@ -301,6 +316,125 @@ pub struct HostFunctionCosts {
+
+    fn add(self, rhs: Self) -> Self::Output {
+        Self {
+            read_value: self.read_value + rhs.read_value,
```
This could use some macro-magic to avoid copy&paste mistakes, or rework of `HostFunctionCosts`.
Not in this PR, though.

alsrdn replied:
I went with the default aversion to using macros here 😁 but it does seem that we use `derive_more` in some other places in the project so I can switch to that for `Add`.

alsrdn replied:
Changed here: https://github.com/casper-network/casper-node/pull/4344/commits/ffe7e18b9a7139defaa1cebf8c97bde61722a229

alsrdn committed:
ee/contract_messages: fix CR comments

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

cspramit commented:
@alsrdn @SaiProServ whats the associated story?

EdHastingsCasperAssociation commented on execution_engine/src/runtime_context/mod.rs:
```diff
@@ -126,6 +127,7 @@ where
             transfers,
             remaining_spending_limit,
             package_kind,
+            last_message_cost: U512::from(0),
```
i don't think the `last_` prefix is necessary

alsrdn replied:
renamed here: https://github.com/casper-network/casper-node/commit/8171806b24ac2178a16b6721587d073eeb42a0d5

alsrdn committed:
types/contract_messages: add getters for `Message` fields

Also add convenience methods to convert contract message related stored
values to their inner types.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
ee/contract_messages: always charge for emit_message

Don't increase the cost if the message was not emitted.
Also simplify config emit message cost config.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

rafal-ch reviewed (approved):
LGMT :+1:

alsrdn commented:
bors r+
